File: santree/core/git.py
"""Git operations for worktree management."""

import subprocess
from dataclasses import dataclass
from pathlib import Path
from typing import List, Optional, Tuple

from santree.core.config import get_santree_dir, get_worktrees_dir


@dataclass
class Worktree:
    """Represents a git worktree."""

    path: Path
    branch: str = ""
    commit: str = ""
    is_bare: bool = False
    is_main: bool = False


class GitOperations:
    """Handle all git worktree operations."""

    def __init__(self, repo_root: Path):
        self.repo_root = repo_root
        self.santree_dir = get_santree_dir(repo_root)
        self.worktrees_dir = get_worktrees_dir(repo_root)

    def detect_default_branch(self) -> str:
        """Detect the default branch (main or master)."""
        # Try remote HEAD first
        result = self._run_git("symbolic-ref", "refs/remotes/origin/HEAD")
        if result.returncode == 0:
            ref = result.stdout.strip()
            return ref.replace("refs/remotes/origin/", "")

        # Fall back to checking if main/master exists
        for branch in ["main", "master"]:
            result = self._run_git("rev-parse", "--verify", f"refs/heads/{branch}")
            if result.returncode == 0:
                return branch

        return "main"  # Default assumption

    def pull_latest(self, branch: str) -> Tuple[bool, str]:
        """Pull latest changes for a branch."""
        # Fetch first
        fetch_result = self._run_git("fetch", "origin", branch)
        if fetch_result.returncode != 0:
            return False, f"Failed to fetch: {fetch_result.stderr}"

        return True, "Successfully fetched latest changes"

    def create_worktree(self, branch_name: str, base_branch: str) -> Tuple[bool, str]:
        """Create a new worktree with a new branch."""
        worktree_path = self.worktrees_dir / branch_name

        # Check if worktree already exists
        if worktree_path.exists():
            return False, f"Worktree already exists at {worktree_path}"

        # Check if branch already exists
        result = self._run_git("rev-parse", "--verify", f"refs/heads/{branch_name}")
        if result.returncode == 0:
            return False, f"Branch '{branch_name}' already exists"

        # Ensure .santree/worktrees directory exists
        self.worktrees_dir.mkdir(parents=True, exist_ok=True)

        # Create worktree with new branch
        result = self._run_git(
            "worktree",
            "add",
            "-b",
            branch_name,
            str(worktree_path),
            base_branch,
        )

        if result.returncode != 0:
            return False, f"Failed to create worktree: {result.stderr}"

        return True, str(worktree_path)

    def list_worktrees(self) -> List[Worktree]:
        """List all worktrees."""
        result = self._run_git("worktree", "list", "--porcelain")
        if result.returncode != 0:
            return []

        worktrees = []
        current_wt: dict = {}

        for line in result.stdout.strip().split("\n"):
            if line.startswith("worktree "):
                current_wt = {"path": Path(line.replace("worktree ", ""))}
            elif line.startswith("HEAD "):
                current_wt["commit"] = line.replace("HEAD ", "")[:8]
            elif line.startswith("branch "):
                current_wt["branch"] = line.replace("branch refs/heads/", "")
            elif line == "bare":
                current_wt["is_bare"] = True
            elif line == "" and current_wt:
                worktrees.append(Worktree(**current_wt))
                current_wt = {}

        if current_wt:
            worktrees.append(Worktree(**current_wt))

        return worktrees

    def remove_worktree(self, branch_name: str, force: bool = False) -> Tuple[bool, str]:
        """Remove a worktree."""
        worktree_path = self.worktrees_dir / branch_name

        if not worktree_path.exists():
            return False, f"Worktree not found: {branch_name}"

        args = ["worktree", "remove"]
        if force:
            args.append("--force")
        args.append(str(worktree_path))

        result = self._run_git(*args)
        if result.returncode != 0:
            return False, f"Failed to remove worktree: {result.stderr}"

        # Delete the branch too
        delete_flag = "-D" if force else "-d"
        branch_result = self._run_git("branch", delete_flag, branch_name)
        if branch_result.returncode != 0:
            # Don't fail if branch deletion fails, just warn
            return True, f"Removed worktree but could not delete branch: {branch_result.stderr}"

        return True, f"Removed worktree and branch: {branch_name}"

    def get_worktree_path(self, branch_name: str) -> Optional[Path]:
        """Get path to a specific worktree."""
        worktree_path = self.worktrees_dir / branch_name
        if worktree_path.exists():
            return worktree_path
        return None

    def _run_git(self, *args) -> subprocess.CompletedProcess:
        """Run a git command in the repo root."""
        return subprocess.run(
            ["git", *args],
            cwd=self.repo_root,
            capture_output=True,
            text=True,
        )

File: santree/source/lib/git.ts
import { execSync, exec } from "child_process";
import { promisify } from "util";
import * as path from "path";
import * as fs from "fs";

const execAsync = promisify(exec);

export interface Worktree {
	path: string;
	branch: string;
	commit: string;
	isBare: boolean;
}

export function findRepoRoot(): string | null {
	try {
		return execSync("git rev-parse --show-toplevel", {
			encoding: "utf-8",
		}).trim();
	} catch {
		return null;
	}
}

export function findMainRepoRoot(): string | null {
	try {
		const gitCommonDir = execSync("git rev-parse --git-common-dir", {
			encoding: "utf-8",
		}).trim();
		return path.dirname(path.resolve(gitCommonDir));
	} catch {
		return null;
	}
}

export function isInWorktree(): boolean {
	try {
		const gitDir = execSync("git rev-parse --git-dir", {
			encoding: "utf-8",
		}).trim();
		const gitCommonDir = execSync("git rev-parse --git-common-dir", {
			encoding: "utf-8",
		}).trim();
		// If they differ, we're in a worktree
		return path.resolve(gitDir) !== path.resolve(gitCommonDir);
	} catch {
		return false;
	}
}

export function isWorktreePath(wtPath: string): boolean {
	try {
		const gitDir = execSync("git rev-parse --git-dir", {
			encoding: "utf-8",
			cwd: wtPath,
		}).trim();
		const gitCommonDir = execSync("git rev-parse --git-common-dir", {
			encoding: "utf-8",
			cwd: wtPath,
		}).trim();
		// If they differ, it's a worktree (not the main repo)
		return path.resolve(wtPath, gitDir) !== path.resolve(wtPath, gitCommonDir);
	} catch {
		return false;
	}
}

export function getCurrentBranch(): string | null {
	try {
		return execSync("git rev-parse --abbrev-ref HEAD", {
			encoding: "utf-8",
		}).trim();
	} catch {
		return null;
	}
}

export function getDefaultBranch(): string {
	try {
		const ref = execSync("git symbolic-ref refs/remotes/origin/HEAD", {
			encoding: "utf-8",
		}).trim();
		return ref.replace("refs/remotes/origin/", "");
	} catch {
		// Fall back to checking if main/master exists
		for (const branch of ["main", "master"]) {
			try {
				execSync(`git rev-parse --verify refs/heads/${branch}`, {
					stdio: "ignore",
				});
				return branch;
			} catch {
				continue;
			}
		}
		return "main";
	}
}

export function listWorktrees(): Worktree[] {
	try {
		const output = execSync("git worktree list --porcelain", {
			encoding: "utf-8",
		});

		const worktrees: Worktree[] = [];
		let current: Partial<Worktree> = {};

		for (const line of output.split("\n")) {
			if (line.startsWith("worktree ")) {
				current.path = line.replace("worktree ", "");
			} else if (line.startsWith("HEAD ")) {
				current.commit = line.replace("HEAD ", "").slice(0, 8);
			} else if (line.startsWith("branch ")) {
				current.branch = line.replace("branch refs/heads/", "");
			} else if (line === "bare") {
				current.isBare = true;
			} else if (line === "" && current.path) {
				worktrees.push(current as Worktree);
				current = {};
			}
		}

		if (current.path) {
			worktrees.push(current as Worktree);
		}

		return worktrees;
	} catch {
		return [];
	}
}

export function getSantreeDir(repoRoot: string): string {
	return path.join(repoRoot, ".santree");
}

export function getWorktreesDir(repoRoot: string): string {
	return path.join(getSantreeDir(repoRoot), "worktrees");
}

export async function createWorktree(
	branchName: string,
	baseBranch: string,
	repoRoot: string,
): Promise<{ success: boolean; path?: string; error?: string }> {
	const dirName = branchName.replace(/\//g, "__");
	const worktreesDir = getWorktreesDir(repoRoot);
	const worktreePath = path.join(worktreesDir, dirName);

	if (fs.existsSync(worktreePath)) {
		return {
			success: false,
			error: `Worktree already exists at ${worktreePath}`,
		};
	}

	// Ensure worktrees directory exists
	fs.mkdirSync(worktreesDir, { recursive: true });

	// Check if branch exists
	let branchExists = false;
	try {
		execSync(`git rev-parse --verify refs/heads/${branchName}`, {
			cwd: repoRoot,
			stdio: "ignore",
		});
		branchExists = true;
	} catch {
		branchExists = false;
	}

	try {
		if (branchExists) {
			await execAsync(`git worktree add "${worktreePath}" "${branchName}"`, {
				cwd: repoRoot,
			});
		} else {
			await execAsync(
				`git worktree add -b "${branchName}" "${worktreePath}" "${baseBranch}"`,
				{ cwd: repoRoot },
			);
		}

		// Save metadata
		const metadata = {
			branch_name: branchName,
			base_branch: baseBranch,
			created_at: new Date().toISOString(),
		};
		fs.writeFileSync(
			path.join(worktreePath, ".santree_metadata.json"),
			JSON.stringify(metadata, null, 2),
		);

		return { success: true, path: worktreePath };
	} catch (e) {
		return {
			success: false,
			error: e instanceof Error ? e.message : "Unknown error",
		};
	}
}

export async function removeWorktree(
	branchName: string,
	repoRoot: string,
	force = false,
): Promise<{ success: boolean; error?: string }> {
	// Find the worktree by branch name using git's worktree tracking
	const worktreePath = getWorktreePath(branchName);

	if (!worktreePath) {
		return { success: false, error: `Worktree not found: ${branchName}` };
	}

	try {
		const forceFlag = force ? "--force" : "";
		await execAsync(`git worktree remove ${forceFlag} "${worktreePath}"`, {
			cwd: repoRoot,
		});

		// Clean up any remaining files (untracked files, node_modules, etc.)
		// git worktree remove doesn't delete untracked files
		if (fs.existsSync(worktreePath)) {
			// Fix permissions first (node_modules often has restricted perms)
			try {
				execSync(`chmod -R u+w "${worktreePath}"`, { stdio: "ignore" });
			} catch {
				// Ignore chmod errors
			}
			fs.rmSync(worktreePath, { recursive: true, force: true });
		}

		// Also delete the branch
		const deleteFlag = force ? "-D" : "-d";
		try {
			await execAsync(`git branch ${deleteFlag} "${branchName}"`, {
				cwd: repoRoot,
			});
		} catch {
			// Branch deletion failed, but worktree was removed
		}

		return { success: true };
	} catch (e) {
		return {
			success: false,
			error: e instanceof Error ? e.message : "Unknown error",
		};
	}
}

export function extractTicketId(branch: string): string | null {
	const match = branch.match(/([a-zA-Z]+)-(\d+)/);
	if (match) {
		return `${match[1]!.toUpperCase()}-${match[2]}`;
	}
	return null;
}

export function getWorktreePath(branchName: string): string | null {
	const worktrees = listWorktrees();
	const wt = worktrees.find((w) => w.branch === branchName);
	return wt?.path ?? null;
}

export function getWorktreeMetadata(
	worktreePath: string,
): { branch_name?: string; base_branch?: string; created_at?: string } | null {
	const metadataPath = path.join(worktreePath, ".santree_metadata.json");
	if (!fs.existsSync(metadataPath)) {
		return null;
	}
	try {
		return JSON.parse(fs.readFileSync(metadataPath, "utf-8"));
	} catch {
		return null;
	}
}

export function hasUncommittedChanges(): boolean {
	try {
		const output = execSync("git status --porcelain", { encoding: "utf-8" });
		return Boolean(output.trim());
	} catch {
		return false;
	}
}

export function hasStagedChanges(): boolean {
	try {
		execSync("git diff --cached --quiet", { stdio: "ignore" });
		return false;
	} catch {
		return true;
	}
}

export function hasUnstagedChanges(): boolean {
	try {
		// Check for modified files
		try {
			execSync("git diff --quiet", { stdio: "ignore" });
		} catch {
			return true;
		}
		// Check for untracked files
		const output = execSync("git ls-files --others --exclude-standard", {
			encoding: "utf-8",
		});
		return Boolean(output.trim());
	} catch {
		return false;
	}
}

export function getGitStatus(): string {
	try {
		return execSync("git status --short", { encoding: "utf-8" }).trim();
	} catch {
		return "";
	}
}

export function getStagedDiffStat(): string {
	try {
		return execSync("git diff --cached --stat", { encoding: "utf-8" }).trim();
	} catch {
		return "";
	}
}

export function getCommitsBehind(baseBranch: string): number {
	try {
		const output = execSync(`git rev-list --count HEAD..origin/${baseBranch}`, {
			encoding: "utf-8",
		});
		return parseInt(output.trim(), 10) || 0;
	} catch {
		return 0;
	}
}

export function getCommitsAhead(baseBranch: string): number {
	try {
		const output = execSync(`git rev-list --count ${baseBranch}..HEAD`, {
			encoding: "utf-8",
		});
		return parseInt(output.trim(), 10) || 0;
	} catch {
		return 0;
	}
}

export function remoteBranchExists(branchName: string): boolean {
	try {
		const output = execSync(`git ls-remote --heads origin ${branchName}`, {
			encoding: "utf-8",
		});
		return output.includes(branchName);
	} catch {
		return false;
	}
}

export function getUnpushedCommits(branchName: string): number {
	try {
		// Check if remote tracking branch exists
		try {
			execSync(`git rev-parse --verify origin/${branchName}`, {
				stdio: "ignore",
			});
		} catch {
			// No remote branch, count all local commits
			const output = execSync("git rev-list --count HEAD", {
				encoding: "utf-8",
			});
			return parseInt(output.trim(), 10) || 0;
		}

		// Count commits ahead of remote
		const output = execSync(`git rev-list --count origin/${branchName}..HEAD`, {
			encoding: "utf-8",
		});
		return parseInt(output.trim(), 10) || 0;
	} catch {
		return 0;
	}
}

export function pullLatest(
	baseBranch: string,
	repoRoot: string,
): { success: boolean; message: string } {
	try {
		// Fetch from origin
		execSync("git fetch origin", { cwd: repoRoot, stdio: "ignore" });

		// Update the base branch
		execSync(`git checkout ${baseBranch}`, { cwd: repoRoot, stdio: "ignore" });
		execSync(`git pull origin ${baseBranch}`, {
			cwd: repoRoot,
			stdio: "ignore",
		});

		return { success: true, message: "Fetched latest changes" };
	} catch (e) {
		return {
			success: false,
			message: e instanceof Error ? e.message : "Failed to pull latest",
		};
	}
}

export function hasInitScript(repoRoot: string): boolean {
	const initScript = path.join(getSantreeDir(repoRoot), "init.sh");
	return fs.existsSync(initScript);
}

export function getInitScriptPath(repoRoot: string): string {
	return path.join(getSantreeDir(repoRoot), "init.sh");
}

export function getLatestCommitMessage(): string | null {
	try {
		return execSync("git log -1 --format=%s", {
			encoding: "utf-8",
		}).trim();
	} catch {
		return null;
	}
}
